content/blog/2021-05-valuable.md

```diff
@@ -0,0 +1,213 @@
+---
+date: "2021-05-24"
+title: "Announcing Valuable, a library for object-safe value inspection"
+description: "May 24, 2021"
+---
+
+Over the past few weeks, we have been working on [Valuable][crate], a new crate
+that provides [object-safe] value inspection. It is almost ready to publish, so
+I thought I'd write an article to introduce it. The crate offers an object-safe
+trait, [`Valuable`][trait], that allows callers to inspect the contents of the
+value, whether fields, enum variants, or primitives, without knowing its type.
+Initially, we wrote Valuable to support [Tracing]; however, it is helpful in
+several scenarios. Object-safe value inspection is a bit of a mouthful, so let's
+start by looking at [Tracing] and why it is needed there.
+
+[Tracing] is a framework for instrumenting Rust programs to collect structured,
+event-based diagnostic information. Some consider it a structured logging
+framework, and while it can fill that use case, it can do a lot more. For
+example, [Console] aims to become a powerful tool for debugging async Rust
+applications and uses Tracing as its backbone. [Tokio] and other libraries emit
+instrumentation via Tracing. Console aggregates the events into a model of the
+application's execution, enabling the developer to gain insights into bugs and
+other issues.
+
+![Instrumentation -> trait object -> event collection](https://i.imgur.com/YJIrHK2.png)
+
+Instrumented applications emit events with rich, structured data, and collectors
+receive the events. Of course, at compile-time, the instrumented application and
+the event collectors do not know about each other. A trait object bridges the
+instrumentation half with the collection half enabling collectors to register
+themselves dynamically. So, passing rich, structured data from the
+instrumentation half to the collector requires passing it through the trait
+object boundary. Tracing supports this at a minimal level today but does not
+support passing nested data.
+
+Let's look at an actual use case. Given an HTTP service, at the start of an HTTP
+request, we want to emit a tracing event that includes relevant HTTP headers.
+The data may look something like this.
+
+```javascript=
+{
+  user_agent: "Mozilla/4.0 (compatible; MSIE5.01; Windows NT)",
+  host: "www.example.com",
+  content_type: {
+    mime: "text/xml",
+    charset: "utf-8",
+  },
+  accept_encoding: ["gzip", "deflate"],
+}
+```
+
+In the application, a Rust struct stores the headers.
+
+```rust=
+struct Headers {
+    user_agent: String,
+    host: String,
+    content_type: ContentType,
+    accept_encoding: Vec<String>,
+}
+
+struct ContentType {
+    mime: String,
+    charset: String,
+}
+```
+
+We want to pass this data to the event collector, but how? We cannot give a
+reference to the `Headers` struct. We could use a data type like
```

Comment:
Initially I thought the reason you could not pass it was due to it being nested, since you had a few lines earlier written this:

> Tracing supports this at a minimal level today but does not support passing nested data.

Maybe try rewriting this sentence?